server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFastViewFactory.java

```diff
@@ -0,0 +1,169 @@
+                .date(Helpers.getDateFromHeaderOrInternalDateOtherwise(mimeMessage, firstMessageResult))
+                .preview(getPreview(firstMessageResult.getMessageId(), mimeMessage))
+                .build();
+        }
+
+        private Optional<Preview> getPreview(MessageId messageId, Message mimeMessage) {
+            return OptionalUtils.orSuppliers(
+                () -> Optional.ofNullable(fastProjections.get(messageId))
+                    .map(MessageFastViewPrecomputedProperties::getPreview),
+                () -> computePreview(mimeMessage));
+        }
+
+        private Optional<Preview> computePreview(Message mimeMessage) {
+            try {
+                 MessageContentExtractor.MessageContent messageContent = messageFullViewFactory.extractContent(mimeMessage);
+                 return messageFullViewFactory.mainTextContent(messageContent)
+                    .map(Preview::compute);
+            } catch (MailboxException e) {
+                LOGGER.error("cannot extract content of the message {}", mimeMessage.getMessageId());
+                return Optional.empty();
+            }
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageFastViewFactory.class);
+
+    private final BlobManager blobManager;
+    private final MessageIdManager messageIdManager;
+    private final MessageFastViewProjection fastViewProjection;
+    private final MessageFullViewFactory messageFullViewFactory;
+
+    @Inject
+    @VisibleForTesting
+    MessageFastViewFactory(BlobManager blobManager, MessageIdManager messageIdManager, MessageFastViewProjection fastViewProjection,
+                           MessageFullViewFactory messageFullViewFactory) {
+        this.blobManager = blobManager;
+        this.messageIdManager = messageIdManager;
+        this.fastViewProjection = fastViewProjection;
+        this.messageFullViewFactory = messageFullViewFactory;
+    }
+
+    @Override
+    public List<MessageFastView> fromMessageIds(List<MessageId> messageIds, MailboxSession mailboxSession) {
+        return Mono.from(fastViewProjection.retrieve(messageIds))
+            .flatMap(previews -> gatherMessageViews(messageIds, mailboxSession, previews))
+            .subscribeOn(Schedulers.boundedElastic())
+            .block();
+    }
+
+    private Mono<List<MessageFastView>> gatherMessageViews(List<MessageId> messageIds, MailboxSession mailboxSession,
+                                                           Map<MessageId, MessageFastViewPrecomputedProperties> fastProjections) {
+        return Flux.merge(
+                fetch(ImmutableList.copyOf(fastProjections.keySet()), FetchGroup.HEADERS, mailboxSession),
+                fetch(withoutPreviews(messageIds, fastProjections), FetchGroup.FULL_CONTENT, mailboxSession))
+            .collectList()
+            .map(messages -> Helpers.toMessageViews(messages, new FromMessageResultAndPreview(blobManager, messageFullViewFactory, fastProjections)));
```

Comment:
Why are we calling FromMessageResultAndPreview on messages that don't have a preview.

Let me suggest:

```
return Flux.merge(
    fetch(ImmutableList.copyOf(fastProjections.keySet()), FetchGroup.HEADERS, mailboxSession)
            .map(messages -> Helpers.toMessageViews(messages, new FromMessageResultAndPreview(blobManager, messageFullViewFactory, fastProjections)),
    fetch(withoutPreviews(messageIds, fastProjections), FetchGroup.FULL_CONTENT, mailboxSession))
            map(messages -> Helpers.toMessageViews(messages, MessageFullViewFactory::fromMessageResult)))
    .collectList()
```

Or something like this.

Reply:
because the `MessageFullViewFactory::fromMessageResult` return a `MessageFullView` which expose more information than asked

Reply:
There's a way to fix that is creating MessateFastView from a MessageFullView but it doesn't makes profit from computation point of view

Reply:
> because the MessageFullViewFactory::fromMessageResult return a MessageFullView which expose more information than asked

I don't think that's a big deal to be returning textBody + textHtml here.

> There's a way to fix that is creating MessateFastView from a MessageFullView but it doesn't makes profit from computation point of view

Come on, a MessageFullView IS a MessageFastView (as it extends MessageFastView)

Reply:
after the discussion, extra properties are already filtered before being responded to clients.